### ISSUE TITLE
Add option to read passphrase from another program's stdout

### ISSUE THREAD

**Issue by seschwar** [weechat/weechat#141]
`msmtp` and `mpop` have a `passwordeval` option.  You can specify a command to run from who's standard output `msmtp` or `mpop` will read the password.

I would like to see this in WeeChat as well, since it will allow for a couple of things:
- integration with standalone password managers like `pass` and `pwsafe`
- integration with password managers of desktop environments through query programs like `gnome-keyring-query` and `kwalletcli`
- reading the passphrase from a file encrypted with `gpg` and therefore leave the prompting to `gpg-agent`
- spawn graphical password prompts with `kdialog`, `x11-ssh-askpass`, `zentiy` and others

**Comment by shioyama** [comment]
:+1: this would be great to have

**Comment by incertia** [comment]
are people still interested in this feature? i can probably implement this sometime soon.

**Comment by amerlyq** [comment]
Yes, insterested. Also I would appreciate if 'space' password can also be supplied from stdin to skip prompt and keep secure data encrypted.

**Comment by shapeshed** [comment]
This would be a great feature and let people manage credentials how they want to.

**Comment by shapeshed** [comment]
Should this be closed if there are no plans to implement it?

**Comment by flashcode** [comment]
It's still open and can be implemented in future; currently I lack time to implement such features.

**Comment by foudfou** [comment]
In case @incertia is still interested in contributing, @flashcode could you provide him with some guidance so he can progress on #906 ?

**Comment by amerlyq** [comment]
Actually, in case of nickserv we could use something like this to read password from cmd stdout
```
/set irc.server.bitlbee.command /exec -nosh -nobg -nostdin -norc -noln -pipe "/msg nickserv identify" pass irc/$server
```
But it doesn't work for `sasl_password` or SSL `password`.
However, as these fields are evaluated, we could use something like this:
```
${eval:/exec -nosh -nobg -nostdin -norc -noln pass irc/$server}
${exec:-nosh -nobg -nostdin -norc -noln:pass irc/$server}
${exec:pass irc/$serve}
```
Of course, currently it doesn't work.
But this idea is rather portable and can be used anywhere, as @flashcode was concerned in #906.

**Comment by binaryplease** [comment]
Hi,

I would be greatly interested in this feature and have read of other people over on freenode that would like this. In particular I use password-store, having another keyring specific to weechat seems kind of redundant and is currently keeping from being able to put my weechat config in a public dotfiles repo. Also, I would have to set up another sync functionality for the passwords between different machines.

A general option to read input from external programs in the config would be great and possibly usefull in other regards too apart from passwords.

My C programming skills are not very good but I would be willing to give it a try, if someone tells me where to start.

**Comment by montaropdf** [comment]
I am also using `pass`, so being able to "connect" pass and weechat would be great.

**Comment by aperum** [comment]
Maybe this could be integrated with minimal effort into the existing /secure feature. Instead of setting a password with "/secure set" there could be a "/secure setcmd freenode <evaled cmd>".

As, according to docu, the sec.conf file is read at program start this limits external passwords to weechat start time but i think it's a good compromise.

**Comment by alrayyes** [comment]
Using a yubikey here. This feature would be great to have!

**Comment by jstkdng** [comment]
I managed to get this working by using WEECHAT_PASSPHRASE env variable:
```sh
#!/bin/sh
export WEECHAT_PASSPHRASE=$(pass irc/weechat/enc)
weechat
```
though, this could become a security issue in the future, dunno, at least it works.

**Comment by binaryplease** [comment]
> I managed to get this working by using WEECHAT_PASSPHRASE env variable:
> 
> ```shell
> #!/bin/sh
> export WEECHAT_PASSPHRASE=$(pass irc/weechat/enc)
> weechat $@
> ```
> 
> though, this could become a security issue in the future, dunno, at least it works.

But this only reads the passphrase to decrypt weechat's storage. It still requires to save the passwords in there. I was looking for a solution to read something like the SASL password directly from pass.

**Comment by jstkdng** [comment]
@binaryplease 
you said:
>having another keyring specific to weechat seems kind of redundant and is currently keeping from being able to put my weechat config in a public dotfiles repo.

I found another easy way to fix this. I'm currently using [chezmoi](https://github.com/twpayne/chezmoi) and it supports using dotfiles templates, this is my sec.conf
```
#
# weechat -- sec.conf
#
# WARNING: It is NOT recommended to edit this file by hand,
# especially if WeeChat is running.
#
# Use /set or similar command to change settings in WeeChat.
#
# For more info, see: https://weechat.org/doc/quickstart
#

[crypt]
cipher = aes256
hash_algo = sha512
passphrase_file = ""
salt = off

[data]
__passphrase__ = on
freenode = "{{ pass "irc/weechat/freenode" }}"
rizon = "{{ pass "irc/weechat/rizon" }}"
```
When "applying" the dotfiles, chezmoi replaces those curly braces with the values weechat needs, and I thus am able to upload my dotfiles without any secrets.
Still, you need weechat to generate the hashes first, then you can change it to a template.

**Comment by binaryplease** [comment]
Thanks for the input on workarounds, but this is not the issue here. Also, I'm not changing my complete dotfiles workflow to a different tool just to have it render weechat passwords as a workaround that still requires me to generate hashes in the first place.

My dotfiles "manager" is a simple git repo with detached work-tree and a zsh alias. No additional software and does 90% of what a tool like chezmoi will without hassle. More info: https://news.ycombinator.com/item?id=11070797

**Comment by jstkdng** [comment]
Oh well, you do you my friend. Hope my workaround helps anyone that stumbles upon this thread.

**Comment by Eluminae** [comment]
This feature still could be awesome :angel:

**Comment by flashcode** [comment]
Good news, I'm working on this, and my plan is:

* Add new option `sec.crypt.passphrase_command` where you can put any external program that must return the passphrase on its standard output (first line, with no extra characters). If no passphrase is received or if wrong, you must enter it manually. Note that only the passphrase is read from an external program, all secured data is still managed by WeeChat itself and can not be stored in a password manager, this is complicated to do this because this requires an asynchronous way to read it without blocking WeeChat. See comments in PR #1552, that will be closed once this change is merged.
* Display a warning when you change your passphrase with `/secure passphrase` if a command is configured: `Important: an external program is configured to read the passphrase on startup (option sec.crypt.passphrase_command); you must ensure this program returns the new passphrase you just defined`.
* Remove option `sec.crypt.passphrase_file`, as just using a file can be done with a command (like `cat /some/file.txt`). Removing this option has a small impact for users using it: on `/upgrade` passphrase is not lost, but if you quit and restart, the passphrase will be asked, until you configure a command to read the file you had. There will be a note in the release notes about this change.

Example with [password-store](https://www.passwordstore.org/):

    /set sec.crypt.passphrase_command "pass show weechat/passphrase"

This may ask you to unlock your GPG key with another passphrase, possibly with a graphical popup. Not a problem: WeeChat waits for the end of the command (this command is called only on startup).

Last thing: there's a comment asking for a way to bypass the manual prompt of passphrase, I'm thinking about a way to do this as well (this would start WeeChat with encrypted data, that are not usable until you give the passphrase manually).
I don't know if this is still required with the changes I'm doing, @amerlyq ?

I stay tuned if you have other ideas or comments.

**Comment by flashcode** [comment]
Done!
Exactly like described above.

**Comment by amerlyq** [comment]
> a way to bypass the manual prompt of passphrase
> I don't know if this is still required with the changes I'm doing, @amerlyq ?

Hmm... probably not. It's good to have, but not irreplacable.
After all it's almost 3 years already :)

Some trivia for my reasons at that time:
I used PC+laptop at work and PC+laptop at home -- it totals to 4 linux systems to sync.
I didn't sync ~/.weechat folder directly because:
* it contains all options at once -- hard to focus only on things I explicitly changed.
* it constantly changes with weechat updates -- hard to distinguish from my own changes.
* it contains sensitive data I don't wish to publish -- patching ignored parts of configs is a mess on sync conflicts.
* my work PC must have different set of passwords/accounts than home one -- constantly merging git branches is a fuss.

So basically what I did -- I started using long script with all weechat commands, which I run on all systems as the first thing in the morning... together with 100+ similar scripts for other programs, if there are any changes.
Accordingly this batch of scripts must be unblocking and error-out in case of problems and *not* ask something in prompt.
Especially considering we don't need this secure data to run batch of commands to upgrade the weechat state -- and without password it feels more logical.

Of course, in these 3 years I simply given up on no-password variant and always read password from `pass`, so current enhancement will allow me to... slightly simplify my scripts and make them more secure (e.g. no password exposure through /proc ENV).

**Comment by StayPirate** [comment]
For those who want to integrate it with keepassxc I suggest to set `sec.crypt.passphrase_command` to the following command:

`echo 'check the explanation attribute ;)' | secret-tool store --label='dummy-entry' explanation 'Because of quirks in the gnome libsecret API, a dummy entry needs to be stored in order to guarantee that this keyring was properly unlocked. More details at http://crbug.com/660005 and https://github.com/keepassxreboot/keepassxc/issues/4443' && secret-tool search client weechat 2>&1 | grep -E '^secret' | cut -d " " -f3`

You need `secret-tool` installed on your system and you need to create the entry with your your passphrase stored within:
`secret-tool store --label='weechat-passphrase' client weechat`

**Comment by ifohancroft** [comment]
I am sorry to use this ticket for this, but with the feature now implemented, would someone please explain how to configure WeeChat to use the gpg key to encrypt/decrypt the file, as well as how to get the password from gpg-agent (in-case we already use the gpg key for other programs, such as mutt, so we don't have to unlock the key separately for each program)

**Comment by flashcode** [comment]
@ifohancroft: this is a question not related directly to WeeChat, I'm not sure about the command you have to do, but you should ensure that when you have to unlock the key, your passphrase is prompted (either in terminal or graphical, it doesn't matter).
If that's the case, you can call this directly from WeeChat, and WeeChat will wait for the command to finish before starting.

**Comment by ifohancroft** [comment]
@flashcode Thank you!
